Adjacent motif features, and more generic feature generating

dawahs opened the pull request:
Ok, apologies in advance for this big commit. The biggest changes are:

1. generic_feature_generator: has things that both submotif and adjacent motif features use, like feature updating and whatnot; I moved a lot of what was in submotif_feature_generator to this, because it'll be shared by any feature generator
2. combined_feature_generator: a more general way to combine feature generators---hier_motif_feature_generator is a case of this, but a special one with special ways to combine thetas, which is why that theta code got moved there
3. adjacent_motif_feature_generator: pretty much the same as submotif feature generator but with a different way of getting the feature index

I also changed the way we update the feature generator after removing features---we couldn't actually pass in a feature generator with zeroed out features unless we already had a fit no features removed!

After that, there are a lot of minor things, like functions that required motif lengths and other submotif-specific things that I had to rework. A lot of these were functions we don't really use anymore, so they just got nuked. In addition, I noticed a few of the tests were outdated, but just needed to change how we call some of the functions in them. I think this is why the "files changed" count is so high. Sorry!

jjfeng commented on combined_feature_generator.py:
```diff
@@ -0,0 +1,123 @@
+    A generic way to combine feature generators.
+    A hierarchical model is a type of this.
+    """
+    def __init__(self, feat_gen_list, feats_to_remove=[]):
```
can you add definition for `@param feats_to_remove`

jjfeng commented on generic_feature_generator.py:
```diff
@@ -0,0 +1,286 @@
+        Create the feature matrices and feature vector dictionary
+        before any mutations have occurred
+
+        @return ObservedSequenceMutations augmented with a feature matrix and dictionary
```
you are actually returning a sparse matrix.

jjfeng commented on generic_feature_generator.py:
```diff
@@ -0,0 +1,286 @@
+        raise NotImplementedError()
+
+    def _get_mutating_pos_feat_idx(self, pos, seq_with_flanks):
+        """
```
add comment for what this does

jjfeng commented on adjacent_motif_feature_generator.py:
```diff
@@ -0,0 +1,93 @@
+from generic_feature_generator import GenericFeatureGenerator
+
+class AdjacentMotifFeatureGenerator(GenericFeatureGenerator):
+    """
```
can you describe this in more detail?

i think it is a k-mer motif feature generator for when the motif is some distance away.

jjfeng commented on adjacent_motif_feature_generator.py:
```diff
@@ -0,0 +1,93 @@
+
+    def update_feats_after_removing(self, feats_to_remove=[]):
+        """
+        take existing SubmotifGenerator and update it with features to remove
```
do you use SubmotifGenerator at all?

jjfeng replied:
actually do you use `SubmotifGenerator` in this code at all or does this class cover all your use cases?

jjfeng commented on adjacent_motif_feature_generator.py:
```diff
@@ -0,0 +1,93 @@
+
+        self.feature_vec_len = len(self.feature_info_list)
+
+    def print_label_from_info(self, info):
```
maybe you can describe that this function prints things from `feature_info_list`?

jjfeng commented on adjacent_motif_feature_generator.py:
```diff
@@ -0,0 +1,93 @@
+    def print_label_from_info(self, info):
+        return "adjacent motif: %s, distance: %d" % (info[0], info[1])
+
+    def _get_feature_dict_for_region(
```
is this general enough to move to `GenericFeatureGenerator`?

jjfeng commented on generic_feature_generator.py:
```diff
@@ -0,0 +1,286 @@
+        @param pos: mutating position
+        @param end_seq: final nucleotide sequence
+
+        @return a dict with the positions next to the given position and their feature index
```
huh? you dont return a dict... im confused what this function is for. it takes in the initial string, the final string, and returns... basically the final string back?

jjfeng commented on generic_feature_generator.py:
```diff
@@ -0,0 +1,286 @@
+        seq_len = len(seq_str)
+        if do_feat_vec_pos is None:
+            do_feat_vec_pos = range(len(seq_str))
+        # don't generate any feature vector for positions in no_feat_vec_pos since it is not in the risk group
```
whats no_feat_vec_pos

jjfeng commented on generic_feature_generator.py:
```diff
@@ -0,0 +1,286 @@
+import scipy.sparse
```
done reviewing this one!

jjfeng commented on adjacent_motif_feature_generator.py:
```diff
@@ -0,0 +1,93 @@
+        ):
+        """
+        @param motif_len: list of motifs to look for at a distance from mutating position
+        @param distance_to_right_flank_end: distance from mutating position to look for nearby motif
```
why is this called `distance_to_right_flank_end`? i thoughts flanks were at the ends of the sequence

jjfeng commented on adjacent_motif_feature_generator.py:
```diff
@@ -0,0 +1,93 @@
+import numpy as np
```
lgtm

jjfeng commented on combined_feature_generator.py:
```diff
@@ -0,0 +1,123 @@
+import numpy as np
```
lgtm

jjfeng commented on hier_motif_feature_generator.py:
```diff
@@ -39,113 +40,110 @@ def __init__(self, motif_lens=[3,5], feats_to_remove=None, left_motif_flank_len_
-                if motif not in self.motif_dict.keys():
-                    self.motif_dict[motif] = dict()
-                if raw_motif_idx is not None:
-                    self.motif_dict[motif][f.left_motif_flank_len] = raw_motif_idx + self.feat_offsets[i]
```
oh you dont need this motif dict anymore? :D

jjfeng commented on hier_motif_feature_generator.py:
```diff
@@ -1,31 +1,32 @@
-from feature_generator import *
+import itertools
```
lgtm

dawahs commented:
Okay, I moved everything to motif_feature_generator.py since you were right, adjacent_motif_feature_generator and submotif_feature_generator had a lot of redundancy (we didn't need the latter, but needed to update the former to take on both cases).

I added documentation, and changed around some variable names to be more reasonable. Also, getting rid of SubmotifFeatureGenerator made some other code outdated, so a few of those files got changed (fit_shmulate_model.py, simulate_shm_tree.py, etc.) as well as some tests.

Everything looks hunky-dory now though. Also, the point of the PR was to make the submotif feature generator more general, but in doing that it we should be able to use most of this now to add different kinds of features, not just motif-based. This should also be compatible with our previous code that fit hierarchical motif features (we still have the HierarchicalMotifFeatureGenerator class).

jjfeng commented on fit_shmulate_model.py:
```diff
@@ -18,7 +18,7 @@
 from models import ObservedSequenceMutations
 from mcmc_em import MCMC_EM
-from submotif_feature_generator import SubmotifFeatureGenerator
+from motif_feature_generator import MotifFeatureGenerator
```
omg i am so happy its not called submotif anymore

jjfeng commented on adjacent_motif_feature_generator.py:
```diff
@@ -1,93 +0,0 @@
-        feat_dict = dict()
-        start_region_idx = max(position - self.left_update_region, 0)
-        end_region_idx = min(position + self.right_update_region, seq_len - 1)
-        update_positions = range(start_region_idx, position) + range(position + 1, end_region_idx)
```
lol omg is this the lost  `+1`.

jjfeng commented on motif_feature_generator.py:
```diff
@@ -0,0 +1,106 @@
+        feat_idx = self.motif_dict[motif]
+        return feat_idx
+
+    def _get_mutated_seq(self, intermediate_seq, pos, end_seq):
```
i dont get it. why does the motif feat gen need to subclass this?

jjfeng commented on motif_feature_generator.py:
```diff
@@ -0,0 +1,106 @@
+        """
+        return mutate_string(
+            intermediate_seq,
+            pos - self.distance_to_start_of_motif + self.combined_offset,
```
like why aren't we just mutating `pos`? why does `pos` need to be adjusted?

jjfeng replied:
oh i see.... maybe you want to avoid using `- self.distance_to_start_of_motif + self.combined_offset` to get the orig flank length of the sequence. can you just store the flank length of sequences to this class?

jjfeng commented on motif_feature_generator.py:
```diff
@@ -0,0 +1,106 @@
+            motif_len=3,
+            distance_to_start_of_motif=None,
+            feats_to_remove=[],
+            combined_offset=0,
```
whats this?

jjfeng commented on motif_feature_generator.py:
```diff
@@ -0,0 +1,106 @@
+        """
+        Our motifs need offsets, but general feature generators do not use this.
+
+        @param intermediate_seq: initial nucleotide sequence
```
i think this should say `initial nucleotide sequence with flanks`. otherwise this function makes no sense

jjfeng commented on motif_feature_generator.py:
```diff
@@ -0,0 +1,106 @@
+
+        @param intermediate_seq: initial nucleotide sequence
+        @param pos: mutating position
+        @param end_seq: final nucleotide sequence
```
same here. i think this probably has flanks in order for this code to make sense

jjfeng commented on hier_motif_feature_generator.py:
```diff
@@ -77,7 +77,8 @@ def get_possible_motifs_to_targets(self, mask_shape):
-            for i, (motif, mutating_pos) in enumerate(self.feature_info_list):
+            for i, (motif, distance_to_start_of_motif) in enumerate(self.feature_info_list):
+                mutating_pos = -distance_to_start_of_motif
                 mutating_nucleotide = motif[mutating_pos]
```
weird things are going to happen when this is positive vs negative. have you thought through those cases? do you want both the positive and negative values for `distance_to_start_of_motif`?

jjfeng commented:
lgtm
